source/credentials_provider_sts.c

```diff
@@ -697,44 +699,66 @@ AWS_STATIC_STRING_FROM_LITERAL(s_region_config, "region");
  * 1. Check `AWS_DEFAULT_REGION` environment variable
  * 2. check `region` config file property.
  */
-static struct aws_string *s_resolve_region(
-    struct aws_allocator *allocator,
-    const struct aws_credentials_provider_sts_options *options) {
+static struct aws_string *s_resolve_region(struct aws_allocator *allocator, const struct aws_profile *profile) {
     /* check environment variable first */
     struct aws_string *region = aws_credentials_provider_resolve_region_from_env(allocator);
     if (region != NULL && region->len > 0) {
         return region;
     }
 
+    if (profile) {
+        const struct aws_profile_property *property = aws_profile_get_property(profile, s_region_config);
+        if (property) {
+            region = aws_string_new_from_string(allocator, aws_profile_property_get_value(property));
+        }
+    }
+
+    return region;
+}
+
+/*
+ * Try to construct an endpoint for the sts service
+ */
+void s_resolve_endpoint(
+    struct aws_allocator *allocator,
+    const struct aws_credentials_provider_sts_options *options,
+    struct aws_string **out_endpoint,
+    struct aws_string **out_region) {
+
     struct aws_profile_collection *profile_collection = NULL;
     struct aws_string *profile_name = NULL;
-
+    const struct aws_profile *profile = NULL;
     /* check the config file */
     if (options->profile_collection_cached) {
         profile_collection = aws_profile_collection_acquire(options->profile_collection_cached);
     } else {
         profile_collection =
             aws_load_profile_collection_from_config_file(allocator, options->config_file_name_override);
     }
-    if (!profile_collection) {
-        goto cleanup;
-    }
-    profile_name = aws_get_profile_name(allocator, &options->profile_name_override);
-    if (!profile_name) {
-        goto cleanup;
+    if (profile_collection) {
+        profile_name = aws_get_profile_name(allocator, &options->profile_name_override);
+        if (profile_name) {
+            profile = aws_profile_collection_get_profile(profile_collection, profile_name);
+        }
     }
-    const struct aws_profile *profile = aws_profile_collection_get_profile(profile_collection, profile_name);
-    if (!profile) {
+
+    *out_region = s_resolve_region(allocator, profile);
+
+    if (aws_credentials_provider_construct_endpoint(
+            allocator,
+            out_endpoint,
+            *out_region,
```

Comment:
code clarity: `*out_region` could be NULL at this point....
and I see that `aws_credentials_provider_construct_endpoint()` only requires region *sometimes*

if region isn't always required, add a comment here to that effect

but if it is always required, maybe early-out if s_resolve_region() returns NULL